components/edit-collective/sections/SetupTwoFactorAuth.js

```diff
@@ -0,0 +1,335 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+import { graphql } from '@apollo/react-hoc';
+import { CheckCircle } from '@styled-icons/fa-solid/CheckCircle';
+import { Info } from '@styled-icons/feather/Info';
+import { Field, Form, Formik } from 'formik';
+import QRCode from 'qrcode.react';
+import { defineMessages, FormattedMessage, injectIntl } from 'react-intl';
+import speakeasy from 'speakeasy';
+
+import { confettiFireworks } from '../../../lib/confettis';
+import { getErrorFromGraphqlException } from '../../../lib/errors';
+import { API_V2_CONTEXT, gqlV2 } from '../../../lib/graphql/helpers';
+import { compose } from '../../../lib/utils';
+
+import Container from '../../Container';
+import { Box, Flex } from '../../Grid';
+import Loading from '../../Loading';
+import LoadingPlaceholder from '../../LoadingPlaceholder';
+import MessageBox from '../../MessageBox';
+import StyledButton from '../../StyledButton';
+import StyledInput from '../../StyledInput';
+import StyledInputField from '../../StyledInputField';
+import StyledTooltip from '../../StyledTooltip';
+import { H2, H3, P } from '../../Text';
+import { withUser } from '../../UserProvider';
+
+const messages = defineMessages({
+  errorWrongLength: {
+    id: 'TwoFactorAuth.Setup.Form.ErrorLength',
+    defaultMessage: 'Incorrect code length. Please re-enter your 6-digit code.',
+  },
+  errorWrongFormat: {
+    id: 'TwoFactorAuth.Setup.Form.ErrorFormat',
+    defaultMessage: 'Incorrect code format. Please enter only numbers.',
+  },
+  inputLabel: {
+    id: 'TwoFactorAuth.Setup.Form.InputLabel',
+    defaultMessage: 'Please enter your 6-digit code without any dashes.',
+  },
+});
+
+const content = () => (
+  <div>
+    <P fontSize="Caption" lineHeight="Caption">
+      <FormattedMessage
+        id="TwoFactorAuth.Setup.AppInfo"
+        defaultMessage="You can use apps such as Google Authenticator, Authy, 1Password, LastPass, or Microsoft Authenticator to scan the QR code and receive a code to input."
+      />
+    </P>
+  </div>
+);
+
+class SetupTwoFactorAuth extends React.Component {
+  static propTypes = {
+    /** From intl */
+    intl: PropTypes.object.isRequired,
+    /** From graphql query */
+    addTwoFactorAuthToAccount: PropTypes.func.isRequired,
+    data: PropTypes.object,
+    /** From parent component */
+    slug: PropTypes.string,
+    collectiveName: PropTypes.string,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      error: null,
+      tokenAdded: null,
+      submitting: false,
+    };
+
+    this.submit = this.submit.bind(this);
+  }
+
+  componentDidMount() {
+    const options = {
+      name: `${this.props.collectiveName}'s Open Collective`,
+      length: 64,
+    };
+    const secret = speakeasy.generateSecret(options);
+    // speakeasy options object does not add issuer as expected so we add it ourselves to the otp url
+    const issuer = '&issuer=Open%20Collective';
```

Comment:
You can use the `URL` API to cleanly do such manipulation.

Reply:
Not possible.

Using both the `universal-url` and the plain `URLSearchParams` API, we get invalid OTP auth URL's because the spaces are encoded as `+` instead of `%20` like they should be.

```
const urlParams = new URLSearchParams({
      issuer: 'Open Collective',
    });
```

produces

![Screenshot 2020-06-30 at 14 37 15](https://user-images.githubusercontent.com/37520401/86132895-77497e80-badf-11ea-85ec-bfd2b7085d3e.png)


```
const urlParams = new URLSearchParams({
      issuer: 'Open%20Collective',
    });
```

produces

![Screenshot 2020-06-30 at 14 37 28](https://user-images.githubusercontent.com/37520401/86132918-7f092300-badf-11ea-9d69-778544615451.png)

Both of which are rejected by the authenticator app.

Reply:
Uhm, I'm surprised. How are yo converting `urlParams` to string?

Reply:
urlParams.toString(), as used elsewhere on the frontend

Reply:
Just digged a bit and the software complaining about the `+` is at fault. Anyway, would be interesting to add a comment explaining why we need to do this, future developers might be tempted to refactor it and fail.

Reply:
Definitely will do!